docs/tutorials/data_provider.md

```diff
@@ -61,3 +61,100 @@ impl AdditiveIdentity {
     }
 }
 ```
+
+## Caching Data Provider
+
+It is not generally required to implement a cache, especially if using the BakedDataProvider. However, if you wish to implement a cache in your data pipeline, you can do so like this:
+
+```rust
+use icu_provider::hello_world::HelloWorldFormatter;
+use icu_provider::prelude::*;
+use icu::locid::locale;
+use lru::LruCache;
+use std::convert::TryInto;
+use std::sync::Mutex;
+use yoke::trait_hack::YokeTraitHack;
+use yoke::Yokeable;
+use zerofrom::ZeroFrom;
+
+pub struct MyLruDataCache<P> {
+    cache: Mutex<LruCache<(DataKey, DataLocale), AnyResponse>>,
+    provider: P,
+}
+
+impl<M, P> DataProvider<M> for MyLruDataCache<P>
+where
+    M: KeyedDataMarker + 'static,
+    M::Yokeable: ZeroFrom<'static, M::Yokeable>,
+    M::Yokeable: icu_provider::MaybeSendSync,
+    for<'a> YokeTraitHack<<M::Yokeable as Yokeable<'a>>::Output>: Clone,
+    P: DataProvider<M>,
+{
+    fn load(&self, req: DataRequest) -> Result<DataResponse<M>, DataError> {
+        let cache_key = (M::KEY, req.locale.clone());
+        {
+            let mut cache = self.cache.lock().unwrap();
+            if let Some(any_resp) = cache.get(&cache_key) {
+                return any_resp.clone_downcast();
```

Comment:
Does this need to be a new method? Why not just call `.clone().downcast()` here? It's another API we will have to maintain.

Reply:
I was having trouble getting the compiler to type infer the right things so I just made it a new method.

Reply:
I still don't like adding two methods to our API just for an example (which I think is fundamentally flawed).

Reply:
I think the methods are useful on their own; it's common in Rust to have two versions of a method, one that works on a move and the other on a borrow. Yoke has `map_project` and `map_project_cloned`; Vec has `.iter()` and `.into_iter()`; and so forth.